Show Identifiers section

kshepherd opened the pull request:
_Attribution: Donated by The Library Code with the support of Technische Universität Hamburg and Technische Informationsbibliothek_

## References
* Requires DSpace/DSpace#8423

## Description
This pull request adds the Angular UI implementation of DSpace/DSpace#8423

List of changes in this PR:
* Model and component for Show Identifiers section
* i18n text for template
* Unit tests for component

For test instructions and notes, see DSpace/DSpace#8423

## Checklist

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR doesn't introduce circular dependencies
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).

tdonohue commented:
@kshepherd : This PR is also failing tests & it has LGTM alerts needing resolution.  Could you see if they are easy to resolve?  Same thing here...we are starting to run out of time to get these reviewed in time for 7.4

kshepherd commented:
@tdonohue there are some tests still failing with this and the DSpace PR, as well as some possible tidy-up needed to the way I wired up request GET and POST methods for the /identifiers path + link. Some devs from 4Science are helping out by looking at this (both PRs).
I appreciate this is a big change and time is running short for 7.4. If a reschedule is needed to meet deadlines for priority 7.4 release tasks then that is completely understandable.

Documentation to help review / test and forming draft of 7.x doco:
https://wiki.lyrasis.org/display/~kshepherd/Identifier+submission+section%2C+pre-registration

kshepherd commented:
The 2 failed 16.x tests are related to failed e2e tests, a connection timeout trying to get search or mydspace config for workspaceitems. I don't know how this would be related, or if there's a way for me to easily trace any backend problems. Any tips?

tdonohue reviewed (changes requested):
@kshepherd : I gave this a test today alongside the backend PR.  Unfortunately though, I'm not able to test DOI creation/minting as I don't have a test DOI service that I can use.  So, I only tested scenarios having to do with Handles & gave a basic review of the code itself.

Overall, the code looks reasonable, though we need a second tester who can verify the DOI features are working.

However, I did notice that the behavior when DOIs are *disabled* isn't ideal. In that scenario, the Identifier section displays this (if I have `identifiers.submission.register = true` on the backend):
![identifiers](https://user-images.githubusercontent.com/483997/213030631-5361023a-563d-4011-b7d9-afaab55bb113.png)

It's awesome that I can pre-register Handles already, but obviously this section should NOT display anything related to DOIs unless the system is configured to use DOIs.  

Beyond that, I found the basics work well.  More inline comments below.

kshepherd commented on src/app/core/data/identifier-data.service.ts:
```diff
@@ -0,0 +1,58 @@
+
+@Injectable()
+@dataService(IDENTIFIERS)
+export class IdentifierDataService extends BaseDataService<IdentifierData> {
```
I added a comment description at the top along the lines of other data services I saw

kshepherd commented on src/app/item-page/edit-item-page/item-status/item-status.component.ts:
```diff
@@ -92,27 +147,66 @@ export class ItemStatusComponent implements OnInit {
+              if (hasValue(identifier) && identifier.identifierType === 'doi') {
+                // The item has some kind of DOI
+                no_doi = false;
+                if (identifier.identifierStatus === '10' || identifier.identifierStatus === '11'
```
I've added a simple array to help translate the status into a string that is the same as the constant variable name used in the backend identifier code:
```
    public String[] statusText = {
        "UNKNOWN",                      // 0
        "TO_BE_REGISTERED",             // 1
        "TO_BE_RESERVED",               // 2
        "IS_REGISTERED",                // 3
        "IS_RESERVED",                  // 4
        "UPDATE_RESERVED",              // 5
        "UPDATE_REGISTERED",            // 6
        "UPDATE_BEFORE_REGISTRATION",   // 7
        "TO_BE_DELETED",                // 8
        "DELETED",                      // 9
        "PENDING",                      // 10
        "MINTED",                       // 11
    };
```

So at the time of constructing the HAL resource the string will populate here instead of the integer.

kshepherd commented on src/app/shared/object-list/identifier-data/identifier-data.component.ts:
```diff
@@ -0,0 +1,56 @@
+  /**
+   * Whether to show the access status badge or not
+   */
+  showAccessStatus: boolean;
```
yes, a few access status reerences here from code i used as a template for the component, i've fixed this up

kshepherd commented on src/app/shared/object-list/identifier-data/identifier-data.component.ts:
```diff
@@ -0,0 +1,56 @@
+/**
+ * Component rendering the access status of an item as a badge
+ */
+export class IdentifierDataComponent {
```
yes, a few access status reerences here from code i used as a template for the component, i've fixed this up

kshepherd commented on src/app/shared/object-list/identifier-data/identifier.model.ts:
```diff
@@ -0,0 +1,12 @@
+import {autoserialize} from 'cerialize';
+
+export class Identifier {
```
done

kshepherd commented on src/app/submission/sections/identifiers/section-identifiers.component.html:
```diff
@@ -0,0 +1,35 @@
+    </ng-container>
+    <!-- DOIs -->
+    <ng-container *ngIf="!(identifierData && identifierData.doi)">
+      <li>{{'submission.sections.identifiers.no_doi' | translate}}</li>
```
i hadn't considered this.. that the "no minted" message is essentially just annoying and irrelevant if DOIs are not configured. I'll think about the best way around this... we can expose backend config to help with this though it's only spring right now. Or, maybe better to have a new config prop in identifiers.cfg that allows finer grained control of what will be displayed: doi, handle, other, or any combo of those

kshepherd commented on src/app/item-page/edit-item-page/item-register-doi/item-registerdoi.component.ts:
```diff
@@ -0,0 +1,94 @@
+import { first, map } from 'rxjs/operators';
+import { hasValue } from '../../../shared/empty.util';
+import { Observable } from 'rxjs';
+import {getItemEditRoute, getItemPageRoute} from '../../item-page-routing-paths';
```
Thanks, I will continue working on any test failures and lint tests as I update the code to address your feedback

kshepherd commented:
@tdonohue I believe I have addressed your concerns here:

1. As per REST Contract discussion, the data model for the identifiers returned in the identifiers REST endpoint has been made more uniform, and matches the submission section data model for identifiers, too
2. Simplified display means an identifier will only be displayed in the Submission Section if it exists. There is no longer any placeholder "no DOI has been..." message in either case.
3. The "Register DOI" button now performs a more generic 'registerIdentifier' operation which takes a required query parameter 'type'. Currently only type=doi is implemented
4. Unit and integration tests are updated to fix any problems and support this behaviour

atarix83 commented on src/app/item-page/edit-item-page/item-status/item-status.component.ts:
```diff
@@ -92,27 +144,66 @@ export class ItemStatusComponent implements OnInit {
+            }
+          }),
+          toArray()
+        ).subscribe((ops) => this.operations$.next(ops));
```
Probably we should refactor this part of code avoiding to have nested subscription

pnbecker replied:
@atarix83 can you please give use more details about how to refactor this? We thought this is a good way to do this.

kshepherd replied:
i believe this also follows the same pattern as the version currently in main, but if there are optimisation or nesting issues that we need to address, happy to look at it - I might just need some further info or pointers from the angular dev perspective.

As per main https://github.com/DSpace/dspace-angular/blob/main/src/app/item-page/edit-item-page/item-status/item-status.component.ts:
```js
      observableFrom(operations).pipe(
        mergeMap((operation) => {
          if (hasValue(operation.featureID)) {
            return this.authorizationService.isAuthorized(operation.featureID, item.self).pipe(
              distinctUntilChanged(),
              map((authorized) => new ItemOperation(operation.operationKey, operation.operationUrl, operation.featureID, !authorized, authorized))
            );
          } else {
            return [operation];
          }
        }),
        toArray()
      ).subscribe((ops) => this.operations$.next(ops));
```

tdonohue replied:
@kshepherd : To provide a hint here... a "nested subscription" is when you have a "subscribe" inside another "subscribe".  This subscribe (line 197) is nested under another subscribe (line 177).  The example code you posted in your comment does *not* appear to have this same issue... it's an example of the better way to code this.

I'll admit, I'm a newbie myself when it comes to RxJS, but there are docs out there about the dangers of nested subscriptions.  Here's a few:
https://blog.brecht.io/rxjs-best-practices-in-angular/#avoiding-nested-subscribes
https://vocon-it.com/2018/09/04/how-to-get-rid-of-nested-subscriptions-using-rxjs-flatmap-and-switchmap/

I'm not the right person to provide advice on how exactly to refactor this yet (I'm still learning too), but I'm hoping this will provide you with the hints you need to work it out.

kshepherd replied:
thanks a lot @artlowel ! I'll give it a try today and let you know how i go.

kshepherd replied:
@artlowel The code compiles OK, and I can see the expected things happening in the map callbacks, but the subscribe never catches anything emitted... it seems that `toArray()` is never reached, which could be some issue where the source observable is never properly completed, from what I can see... I will keep digging and look into how I can debug the whole subscription pipeline

artlowel replied:
`toArray()` only emits if all the observables before have completed, so perhaps you need a `take(1)` at the end of the `this.authorizationService.isAuthorized` pipe to ensure it completes? That would be surprising though as the original code used `toArray()` too, but it's worth a shot.

Something to help debug could be to add a `tap((v) => console.log('something to identify the log', v)),` at any place in the pipe you want log what's passing by at that point.

kshepherd replied:
@artlowel thanks. After some debugging with tap, and different pipe structures, I settled on something which I think does the job well and simplifies some of the logic, as well as avoiding nested subscriptions. I think part of the problem with the toArray / emit completion was the way I was reading the REST configuration property.

kshepherd commented:
@atarix83 I have pushed changes that tidy up the names, and also use the new endpoints as per the DSpace REST refactor for this PR. I have not successfully rewritten the nested subscription code yet. I think I know the technique I want to use (subscribe to showRegister$ so that I can detect changes in that logic, and essentially call .next() on operations) but just the way it's originally written, where the auth checks really weren't designed to be read more than once (i think?) is making it tricky for me to design properly.
Should have more progress on that soon, I hope. Any tips are welcome.

I expect to see some lint / test errors as I haven't fixed all of those yet, I also have some comments / doc to add. Pushing what I have now just to show progress,

tdonohue reviewed (approved):
👍 Thanks @kshepherd .  I've tested this again today with **Handles only** and it works well to pre-register Handles.  However, I don't have a way to test this with DOIs, so I'm hoping that @atarix83 (or someone else) can test the entire DOI process... as I cannot tell if it's working.  

The code though looks reasonable, so I'll go ahead and submit a +1.  That said, I don't think we should merge this until someone has tested the DOI process.

kshepherd commented on src/app/item-page/edit-item-page/item-status/item-status.component.ts:
```diff
@@ -72,12 +91,37 @@ export class ItemStatusComponent implements OnInit {
+      );
+
+      // Observable for configuration determining whether the Register DOI feature is enabled
+      let registerConfigEnabled$: Observable<boolean> = this.configurationService.findByPropertyName('identifiers.item-status.register-doi').pipe(
```
@atarix83 I request the property to determine if the button should be drawn at all, then in the pipe where features are checked, the DOI button is added first (if applicable), so it is also checked as a feature (just like all the other operations in the array) to set the enabled/disabled state.

I wanted these to both apply, so that people who chose not to use this new functionality at all weren't bothered by an irrelevant button. If the config is enabled but the user lacks authorization for some reason or other, then it will appear but be disabled.

Does that make sense?

atarix83 replied:
it makes sense, thanks

pnbecker commented:
@tdonohue If @atarix83 agrees that @kshepherd explaination above is acceptable for him this PR is ready to be merged. As mentioned in the backend, I tested it again without and with DOIs activated. It works like a charm.

mirkoscherf commented:
"curation-task.task.registerdoi.label" has been changed here to "curation-task.task.register-doi.label" and now the label text is not displayed properly as can be seen on demo7.dspace.org.
![image](https://github.com/DSpace/dspace-angular/assets/45286669/0104b59f-bb83-4a33-b9db-a653353dd0b1)

mirkoscherf commented:
@tdonohue Alright, will do!
